### PR TITLE
iqcc ilc solver fork

### PR DESCRIPTION
**Pull request by MPCoons**

### CODE REVIEW

File: tangelo/toolboxes/ansatz_generator/_qubit_ilc.py [hunk 1]
```diff
@@ -234,12 +253,87 @@ def get_ilc_params_by_diag(qubit_ham, ilc_gens, qmf_var_params):
...
-        denom_sum += pow(gs_coefs[i].real, 2.) + pow(gs_coefs[i].imag, 2.)
+        denom_sum += abs(gs_coefs[i])**2
         beta = np.arcsin(gs_coefs[i] / np.sqrt(denom_sum))
         ilc_var_params.append(beta.real)
     del ilc_gens[0]
     return ilc_var_params
+
+
+def build_ilc_qubit_op_list(acs_gens, ilc_params):
+    """Returns a list of 2N - 1 ILC generators to facilate generation of a circuit
```
Comment (AlexandreF-1qbit) [inline tangelo/toolboxes/ansatz_generator/_qubit_ilc.py:268]: facilitate

File: tangelo/toolboxes/ansatz_generator/_qubit_ilc.py [hunk 2]
```diff
@@ -234,12 +253,87 @@ def get_ilc_params_by_diag(qubit_ham, ilc_gens, qmf_var_params):
...
+        ilc_op_list.append(-.5 * ilc_params[i] * acs_gens[i])
+    ilc_op_list.append(ilc_params[0] * acs_gens[0])
+    for i in range(1, n_amps):
+        ilc_op_list.append(-.5 * ilc_params[i] * acs_gens[i])
+    return ilc_op_list
+
+
+def ilc_op_dress(qubit_op, ilc_gens, ilc_params):
+    """Performs transformation of a qubit operator with the ACS of ILC generators and
+    parameters. For a set of N generators, each qubiti operator transformation results
```
Comment (AlexandreF-1qbit) [inline tangelo/toolboxes/ansatz_generator/_qubit_ilc.py:292]: qubit

**Review by AlexandreF-1qbit — COMMENTED** [review]
Only small comments (mostly typos). I focused on the unittests; I did not see something missing and I think there is a good code overage. I let @JamesB-1qbit and @ValentinS4t1qbit approve the general design since I am a bit out of the loop on this.

**Review by JamesB-1qbit — CHANGES_REQUESTED** [review]
Looks good to me. One small suggestion.

**Review by JamesB-1qbit — APPROVED** [review]

File: tangelo/toolboxes/ansatz_generator/ilc.py [hunk 3]
```diff
@@ -87,50 +100,38 @@ def __init__(self, molecule, mapping="jw", up_then_down=False, acs=None,
...
-        if qubit_ham is None:
+        if not qubit_ham:
             self.fermi_ham = self.molecule.fermionic_hamiltonian
             self.qubit_ham = fermion_to_qubit_mapping(self.fermi_ham, self.mapping,
                                                       self.n_spinorbitals, self.n_electrons,
                                                       self.up_then_down, self.spin)
 
-        self.qmf_var_params = qmf_var_params
-        if self.qmf_var_params is None:
+        if isinstance(qmf_var_params, list):
```
Comment (ValentinS4t1qbit) [inline tangelo/toolboxes/ansatz_generator/ilc.py:114]: `self.qmf_var_params = np.array(qmf_var_params) if isinstance(qmf_var_params, list) else qmf_var_params`

File: tangelo/toolboxes/ansatz_generator/qcc.py [hunk 4]
```diff
@@ -108,12 +115,13 @@ def __init__(self, molecule, mapping="jw", up_then_down=False, dis=None,
                                                       self.n_spinorbitals, self.n_electrons,
                                                       self.up_then_down, self.spin)
 
-        self.qmf_var_params = qmf_var_params
-        if not self.qmf_var_params:
+        if isinstance(qmf_var_params, list):
```
Comment (ValentinS4t1qbit) [inline tangelo/toolboxes/ansatz_generator/qcc.py:118]: see previous comment

File: tangelo/toolboxes/ansatz_generator/tests/test_ilc.py [hunk 5]
```diff
@@ -64,7 +61,7 @@ def test_ilc_incorrect_number_var_params(self):
 
     @staticmethod
     def test_gauss_elim_over_gf2_sqrmat():
-        """ Verify behavior of the Gaussian elimination over the binary field function. """
+        """ Verify behavior of the Gaussian elimination for a square matrix."""
```
Comment (ValentinS4t1qbit) [inline tangelo/toolboxes/ansatz_generator/tests/test_ilc.py:64]: regarding my earlier comment about how [0]*n is a multiple of any line of a matrix with n columns: does it make sense to test / define the behaviour of your code for a matrix that contains at least a line full of 0s?

File: tangelo/toolboxes/operators/operators.py [hunk 6]
```diff
@@ -65,6 +67,24 @@ def frobenius_norm_compression(self, epsilon, n_qubits):
...
+        Args:
+            n_qubits (int): Number of qubits in the register.
+
+        Returns:
+            int: The maximum number of possible qubit Hamiltonian terms.
+        """
+
+        n_terms = 0
+        for i in range(n_qubits // 2):
+            n_terms += comb(n_qubits, 2 * i, exact=True) * 3**(n_qubits - 2 * i)
```
Comment (ValentinS4t1qbit) [inline tangelo/toolboxes/operators/operators.py:85]: i suggest `2 * i` => `2*i` for readability.

**Review by ValentinS4t1qbit — APPROVED** [review]
Good work, a few comments mostly on the A*x = b in Z/2Z

**Review by ValentinS4t1qbit — CHANGES_REQUESTED** [review]
I misclicked earlier, my comments need to be addressed with changes

File: tangelo/toolboxes/ansatz_generator/_qubit_ilc.py [hunk 7]
```diff
@@ -127,62 +130,85 @@ def gauss_elim_over_gf2(a_mat, b_vec=None):
...
-    z_vals, z_sln, piv_idx = [], [-1] * n_cols, 0
-    # check that b_vec was properly supplied; ortherwise initialize as a vector of zeros
     if not isinstance(b_vec, np.ndarray):
         b_vec = np.zeros((n_rows, 1))
+
+    # append the initial solution vector as the last column of a_mat; update n_cols
     a_mat = np.append(a_mat, b_vec, axis=1)
     n_cols += 1
+
+    # check that all elements are positive and either 0 or 1
```
[the remark this answers is not in the document]
Comment (MPCoons) [reply tangelo/toolboxes/ansatz_generator/_qubit_ilc.py:145]: Added a warning message and simplified those lines to one.

File: tangelo/toolboxes/ansatz_generator/_qubit_ilc.py [hunk 8]
```diff
@@ -127,62 +130,85 @@ def gauss_elim_over_gf2(a_mat, b_vec=None):
...
+    a_mat = abs(a_mat)
+    two_mat = np.full((n_rows, n_cols), 2, dtype=int)
+    a_mat = np.fmod(a_mat, two_mat)
+
+    # check for linear dependency of the rows of a_mat
+    linear_dep = False
+    for i in range(n_rows):
+        for j in range(i+1, n_rows):
+            if (a_mat[i][:] == a_mat[j][:]).all():
+                warnings.warn(f"Linear dependency detected: rows {i} and {j} are "
```
[the remark this answers is not in the document]
Comment (MPCoons) [reply tangelo/toolboxes/ansatz_generator/_qubit_ilc.py:155]: True -- in the above comment, we check for duplicates. I will add a check for rows full of 0s as well.

File: tangelo/toolboxes/ansatz_generator/tests/test_ilc.py (hunk 5 above)
Comment (MPCoons) [reply tangelo/toolboxes/ansatz_generator/tests/test_ilc.py:64]: Yes, added another test for the case when there is a matrix with a row of all zeros.

File: tangelo/toolboxes/ansatz_generator/_qubit_ilc.py [hunk 9]
```diff
@@ -127,62 +129,86 @@ def gauss_elim_over_gf2(a_mat, b_vec=None):
...
-    # check that b_vec was properly supplied; ortherwise initialize as a vector of zeros
     if not isinstance(b_vec, np.ndarray):
         b_vec = np.zeros((n_rows, 1))
+
+    # append the initial solution vector as the last column of a_mat; update n_cols
     a_mat = np.append(a_mat, b_vec, axis=1)
     n_cols += 1
+
+    # force all entries of a_mat to be either 0 or 1.
+    warnings.warn("Reducing input matrix elements modulo 2 to create a binary matrix.", RuntimeWarning)
```
Comment (ValentinS4t1qbit) [inline tangelo/toolboxes/ansatz_generator/_qubit_ilc.py:145]: So the default behaviour is that, everytime this function is called, we print a warning on the user screen?

**Review by ValentinS4t1qbit — APPROVED** [review]
It seems everytime the ILC class is instantiated, multiple warning messages coming from the gf2solve will be printed. If this is acceptable behavior according to Marc, and the number of prints is reasonable, I'm ok with that

File: tangelo/toolboxes/ansatz_generator/_qubit_ilc.py (hunk 9 above)
Comment (MPCoons) [reply tangelo/toolboxes/ansatz_generator/_qubit_ilc.py:145]: Yes -- originally I just changed a_mat to a binary matrix without warning the user. I added this warning message based on how I interpreted one of your PR review comments. Would you prefer to do it another way such as adding a verbose flag?

Comment (MPCoons) [reply tangelo/toolboxes/ansatz_generator/_qubit_ilc.py:145]: Changed it so that it only prints a message if a_mat is not passed as a binary matrix.

**Pull request merged**